V almonacid/ch2323/add wallet transfer support

v-almonacid opened the pull request:
This PR removes the wallet check button and adds the shelley wallet upgrade flow and hopefully all  the functions required to generate the wallet migration transaction. ~~However, the transaction construction will fail as there are still some issues that need to be fixed in the rust bindings~~.

UPDATE: 
- tx construction/signing/submission should be working now. 
- TODO in follow-up PR(s):
    - improve UI (add tx id, add rolling wheel during processing, etc)
   - display correct balance once wallet is added

## Restore wallet on Shelley button:
![image](https://user-images.githubusercontent.com/7271744/71212658-84fc8c80-2290-11ea-871b-cc8ffaf988ac.png)

## Modification on restore screen
The restore wallet screen is modified by just passing a `isShelleyWallet` navigation parameter, which changes the navigation theme ad-hoc.
![image](https://user-images.githubusercontent.com/7271744/71212770-c55c0a80-2290-11ea-8f38-66ca2344939a.png)

## Verify restored wallet
![image](https://user-images.githubusercontent.com/7271744/71212881-ff2d1100-2290-11ea-9cfa-6964268d284d.png)

## Wallet upgrade
![image](https://user-images.githubusercontent.com/7271744/71212893-081de280-2291-11ea-927e-93475b06bc18.png)

## Confirm wallet upgrade
![image](https://user-images.githubusercontent.com/7271744/71214835-e4f53200-2294-11ea-84c2-35367f5fdd06.png)

## When no upgrade tx is required
![image](https://user-images.githubusercontent.com/7271744/71214854-f2122100-2294-11ea-9393-b3d7e1fc1c4d.png)

v-almonacid committed:
add yoroiTransfer and some utility functions

v-almonacid committed:
wip

v-almonacid committed:
add yoroiTransfer fx

v-almonacid committed:
first iteration of wallet restore/migrate UI

v-almonacid committed:
nico's and marta's feedback + some fixes and cleanup

v-almonacid committed:
grammar check with leo

v-almonacid committed:
add shelley address derivation and some refactoring

v-almonacid committed:
bump react-native-chain-libs

v-almonacid committed:
connected yoroi transfer tx building to ui

v-almonacid committed:
removed unnecessary file

v-almonacid committed:
send yoroiTransfer tx to api. Bump bindings. Misc fixes

v-almonacid committed:
apply latest changes in utxo transactions from yoroi-frontend

v-almonacid committed:
remove balance check feature

v-almonacid committed:
move restore shelley wallet flow to separate button in wallet init

v-almonacid committed:
fix snapshot date, update shelley backend url and some small fixes/refactoring

small refactoring/fix

fix

update shelley backend url

v-almonacid committed:
fix merge inconsistencies with shelley

v-almonacid committed:
fix merge issues

v-almonacid committed:
use per-certificate fee and update chain libs

SebastienGllmt commented:
https://github.com/Emurgo/yoroi-frontend/pull/1287 make sure this is included also

v-almonacid committed:
fix sendall when small UTXO is included

v-almonacid committed:
misc fixes

at this point the wallet transfer tx seems to be correctly generated, with no error from the bindings.

v-almonacid committed:
some refactoring and small fixes

v-almonacid committed:
attempt to implement shelley wallet creation

This should allow to add a new Shelley wallet to the app and generate group addresses. More testing/debugging is still needed.

v-almonacid committed:
fix tx submission and error handling

SebastienGllmt commented on src/i18n/global-messages.js:
```diff
@@ -109,6 +127,16 @@ export const errorMessages = {
+    },
+    message: {
+      id: 'global.actions.dialogs.insufficientBalance.message',
+      defaultMessage: '!!Not enough money to make this transaction',
```
nit: probably best to avoid the word "money". Prefer things like "balance", "funds" or ADA

SebastienGllmt commented on src/crypto/shelley/transactions/utxoTransactions.js:
```diff
@@ -363,34 +337,44 @@ export const sendAllUnsignedTxFromUtxo = async (
       await fakeIOBuilder.add_input(input)
     }
     await fakeIOBuilder.add_output(
+      // note: currently failing
```
todo?

SebastienGllmt commented on src/crypto/shelley/transactions/utxoTransactions.js:
```diff
@@ -363,34 +337,44 @@ export const sendAllUnsignedTxFromUtxo = async (
+        : await Payload.no_payload()
     const feeValue = await (await fakeIOBuilder.estimate_fee(
       feeAlgorithm,
       // can't add a certificate to a UTXO transaction
```
remove this comment. It's not accurate

SebastienGllmt commented on src/crypto/chain.js:
```diff
@@ -39,6 +45,54 @@ export class AddressGenerator {
   }
 }
 
+// TODO: make ShelleyAddressGenerator a subclass of AddressGenerator
```
TODO?

SebastienGllmt commented on src/crypto/chain.js:
```diff
@@ -39,6 +45,54 @@ export class AddressGenerator {
+  addressChain: Bip32PublicKey
+  // stakingKey: PublicKey
+  // for now, need to store accountKey instead of stakingKey because the latter
+  // is PublicKey instance and from_bech32() is currently mising from the
```
is this still missing?

v-almonacid replied:
Yes, but finally I will just store the account-level public key, as we did in Byron.

SebastienGllmt commented on src/crypto/byron/util.js:
```diff
@@ -9,22 +9,24 @@ import bs58 from 'bs58'
 } from '../errors'
 import {AddressChain, AddressGenerator} from '../chain'
+import {ADDRESS_TYPE_TO_CHANGE} from '../commonUtils'
+// TODO: refactor to remove these imports here as they are causing a cycle
```
TODO?

SebastienGllmt commented on src/components/WalletInit/RestoreWallet/WalletCredentialsScreen.js:
```diff
@@ -1,57 +1,356 @@
+    })
+
+  onCheck = async () => {
+    // TODO: there is currently a bug when displaying the UpgradeCheckModal
```
TODO?

v-almonacid replied:
Still todo but not critical IMHO. I haven't figured out why this happens.

SebastienGllmt commented on src/components/WalletInit/RestoreWallet/WalletCredentialsScreen.js:
```diff
@@ -1,57 +1,356 @@
+  }, 1000)
+
+  onSubmitWalletCredentials = ({name, password}) => {
+    // TODO: make sure this call is executed only once
```
TODO?

v-almonacid committed:
changes after seba's review
